src/components/features/gift-order/RecipientModal.tsx

```diff
@@ -0,0 +1,297 @@
+import React from 'react';
+import type { RefObject } from 'react';
+import styled from '@emotion/styled';
+import { theme } from '@/styles/theme';
+import RecipientFormItem from './RecipientFormItem';
+import { useForm, useFieldArray } from 'react-hook-form';
+import { z } from 'zod';
+import { zodResolver } from '@hookform/resolvers/zod';
+
+type Recipient = { name: string; phone: string; quantity: number };
+type RecipientsForm = { recipients: Recipient[] };
+
+const recipientSchema = z.object({
+  name: z.string().min(1, '이름을 입력해주세요.'),
+  phone: z
+    .string()
+    .min(1, '전화번호를 입력해주세요.')
+    .regex(/^010[0-9]{8}$/, '올바른 전화번호 형식이 아니에요.'),
+  quantity: z.coerce.number().min(1, '구매 수량은 1개 이상이어야 해요.'),
+});
+
+const recipientArraySchema = z
+  .array(recipientSchema)
+  .min(0, '최소 0명')
+  .max(10, '최대 10명')
+  .superRefine((arr, ctx) => {
+    const phoneMap = new Map<string, number[]>();
+    arr.forEach((r, i) => {
+      if (!phoneMap.has(r.phone)) phoneMap.set(r.phone, []);
+      phoneMap.get(r.phone)!.push(i);
```

Comment:
지금 상황은 코드에서 undefined가 될 상황이 아니지만, 조건문에 따른 타입 추론이 명확하게 되지 않기에, 타입으로 단언하신거 같아요!
하지만 보다 타입이 자연스럽게 추론되도록 코드를 작성하면 좋을거 같아요!

```suggestion
const indices = phoneMap.get(r.phone) ?? [];
      indices.push(i);
      phoneMap.set(r.phone, indices);
```